### DISCUSSION TITLE
catchToEffect deprecation without alternative for supported platform

### DISCUSSION THREAD

**Discussion by jaredh159** [pointfreeco/swift-composable-architecture#1987] [Q&A] [answered]
huge fan of TCA and your work. thanks so much for all your libraries. 🙏 

currently working with the `prerelease/1.0` branch, building a macOS app that targets `10.15` - `13`.  `10.15` (Catalina) is [a supported platform](https://github.com/pointfreeco/swift-composable-architecture/blob/prerelease/1.0/Package.swift#L9) according to the `Package.swift`.

but, I'm getting deprecation warnings trying to use `.catchToEffect`, with a suggestion that I iterate over `publisher.values`.  that's all well and good, but `publisher.values` isn't available until macOS 12. so, i can't convert to it.

can this warning/deprecation be scoped to platforms that support the alternative? or is there some alternative I'm not aware of that would work without deprecation/warning as far back as `10.15`?

and I'm hoping you aren't planning on removing this api soon, right? unless I'm missing something that would force a platform compat bump all the way to macos 12.

apologies if there's something obvious i'm overlooking here.

**Comment by stephencelis** [comment, marked as answer]
Hi @jaredh159! Not sure if you've pulled recently, but last week we added to the deprecation message to also suggest using `Effect.publisher`, which acts as a Combine bridge for earlier for older platforms:

https://github.com/pointfreeco/swift-composable-architecture/blob/132cb3bae2d333c75ff2c952086d1dcd53ba392e/Sources/ComposableArchitecture/Effects/Publisher.swift#L383

You can then use `Publisher.catch` and other Combine operations on the publisher in that block:

```swift
return .publisher {
  self.dependencyThatCanFail()
    .map { .response(.success($0)) }
    .catch { .response(.failure($0)) }
}
```

I'm going to convert this to a discussion since I think it's more of a question than a bug.

**Reply by jaredh159** [reply]
ahh, gotcha, thanks. and yup, this does make more sense as a discussion.

no, i haven't pulled recently, so i didn't see the revised deprecation note.  that said, i tried to just now by running `swift package update`, and i'm getting an SPM resolution fatal error:

```
error: Couldn’t check out revision ‘fc17996ded63b136741ab2e0c0e0d549a8486adc’:
fatal: reference is not a tree: fc17996ded63b136741ab2e0c0e0d549a8486adc
```

`fc17996de...` is the sha of the latest commit on the `prerelease/1.0` branch, so it seems like it's failing to pull it. 

is `swift package update` the correct way to update the dependency? I confess this is the first time i've depended on an actively changing branch, so maybe there's some other way to do it...

**Reply by acosmicflamingo** [reply]
SPM is caching something in DerivedData. I have been able to get this error to go away by clearing the following:
```
${DerivedData}/<ProjectDir>/SourcePackages/checkouts/<whatever Composable Architecture is called here>
```

**Reply by jaredh159** [reply]
@acosmicflamingo thanks, that was the tip i needed, a quick `rm -rf .build/` and then `swift package update` worked like a charm. shoulda guessed that.

**Reply by Alex293** [reply]
if you use Xcode you can also use file -> spm -> reset cache to get the same behaviour

**Reply by tlandsman** [reply]
I'm still confused.  Can you give an example of how to use `EffectTask.run`'  or  `EffectTask.publisher` replaces `catchToEffect`?

**Reply by jaredh159** [reply]
@tlandsman the example Stephen posted is basically what I ended up doing, stuff like:

```swift
return .publisher {
  self.dependencyThatCanFail()
    .map { .response(.success($0)) }
    .catch { .response(.failure($0)) }
}
```

or, in my case, something like:

```swift
return .publisher {
  websocket.receive()
    .map { .websocket(.receivedMessage($0)) }
    .receive(on: mainQueue)
}
```

**Reply by tlandsman** [reply]
Thanks! Much appreciated.
